### PR TITLE
[SPARK-8654] [SQL] Analysis exception when using NULL IN (...) : invalid cast

### PR DESCRIPTION
**Pull request by dilipbiswal**
In the analysis phase , while processing the rules for IN predicate, we
compare the in-list types to the lhs expression type and generate
cast operation if necessary. In the case of NULL [NOT] IN expr1 , we end up
generating cast between in list types to NULL like cast (1 as NULL) which
is not a valid cast.

The fix is to find a common type between LHS and RHS expressions and cast
all the expression to the common type.

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

**Comment by dilipbiswal** [comment]
@cloud-fan 
Hi Wenchen, Can you please review the update to the test case ? Thank you.

**Comment by cloud-fan** [comment]
ok to test

**Comment by dilipbiswal** [comment]
@cloud-fan , hi, how do we trigger a retest ?

**Comment by dilipbiswal** [comment]
@cloud-fan 
It seems to fail now with the following exception ..
hudson.plugins.git.GitException: Failed to fetch from https://github.com/apache/spark.git
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:735)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:983)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1016)
	at hudson.scm.SCM.checkout(SCM.java:485)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1277)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:610)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:532)
	at hudson.model.Run.execute(Run.java:1741)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:408)
Caused by: hudson.plugins.git.GitException: Command "git fetch --tags --progress https://github.com/apache/spark.git +refs/pull/9036/*:refs/remotes/origin/pr/9036/*" returned status code 143:
stdout: 
stderr: error: RPC failed; result=18, HTTP code = 200

What do we do here ?

**Comment by cloud-fan** [comment]
you can just conmment "retest this please", and our jenkins will run the test for you :)

**Comment by dilipbiswal** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercionSuite.scala [hunk 1]
```diff
@@ -470,7 +470,8 @@ class HiveTypeCoercionSuite extends PlanTest {
     )
     ruleTest(inConversion,
       In(Literal("a"), Seq(Literal(1), Literal("b"))),
-      In(Literal("a"), Seq(Cast(Literal(1), StringType), Cast(Literal("b"), StringType)))
+      In(Cast(Literal("a"), StringType), Seq(Cast(Literal(1), StringType),
+         Cast(Literal("b"), StringType)))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercionSuite.scala]: nit:
```
In(Cast(Literal("a"), StringType),
  Seq(Cast(Literal(1), StringType), Cast(Literal("b"), StringType)))
```
looks better

**Comment by dilipbiswal** [comment]
@cloud-fan 
Thanks.. Fixed it.

**Comment by dilipbiswal** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 2]
```diff
@@ -304,7 +304,10 @@ object HiveTypeCoercion {
   }
 
   /**
-   * Convert all expressions in in() list to the left operator type
+   * Convert the value and in list expressions to the common operator type
+   * by looking at all the argument types and finding the closest one that
+   * all the arguments can be cast to. When no common operator type is found
+   * an Analysis Exception is raised.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: "the original one will be returned and an Analysis Exception will be raised at type checking phase"
We don't throw exception here right?

Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: @cloud-fan 
You are right Wenchen.I just wanted to somehow mention that an exception will be raised on a data type mistmatch. I will reword it to the way you suggest. ""the original one will be returned and an Analysis Exception will be raised at type checking phase". Let me push a commit.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM pending test.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
@cloud-fan
Hi Wenchen, the test failed which looks unrelated. I wanted to retest .. but looks like its not triggering it. Can you please help ?

**Comment by marmbrus** [comment]
Test this please

**Comment by cloud-fan** [comment]
ok to test

**Comment by dilipbiswal** [comment]
@mambrus
Hi Michael, the tests have passed. Please let me know if anything is pending before it gets merged.
Thanks a lot.

**Comment by dilipbiswal** [comment]
@marmbrus 
Hi Michael,  the tests have passed. Please let me know if anything is pending before it gets merged.
Thanks a lot.

**Comment by dilipbiswal** [comment]
@marmbrus @rxin

**Comment by marmbrus** [comment]
Thanks!  I'll try and fix the conflicts while merging to master.
